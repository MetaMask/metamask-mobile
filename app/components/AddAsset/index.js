import React, { Component } from 'react';
import { SafeAreaView, StyleSheet } from 'react-native';
import { colors, fontStyles } from '../../styles/common';
import DefaultTabBar from 'react-native-scrollable-tab-view/DefaultTabBar';
import AddCustomToken from '../AddCustomToken';
import SearchTokenAutocomplete from '../SearchTokenAutocomplete';
import ScrollableTabView from 'react-native-scrollable-tab-view';
import PropTypes from 'prop-types';
import { strings } from '../../../locales/i18n';
import AddCustomCollectible from '../AddCustomCollectible';
import { getNavigationOptionsTitle } from '../Navbar';

const styles = StyleSheet.create({
	wrapper: {
		flex: 1,
		backgroundColor: colors.white
	},
	tabUnderlineStyle: {
		height: 2,
		backgroundColor: colors.primary
	},
	tabStyle: {
		paddingBottom: 0
	},
	textStyle: {
		fontSize: 16,
		letterSpacing: 0.5,
		...fontStyles.bold
	}
});

/**
 * Component that provides ability to add assets.
 */
export default class AddAsset extends Component {
	static navigationOptions = () => getNavigationOptionsTitle(strings('add_asset.title'));

	state = {
		address: '',
		symbol: '',
		decimals: ''
	};

	static propTypes = {
		/**
		/* navigation object required to push new views
		*/
		navigation: PropTypes.object
	};

	renderTabBar() {
		return (
			<DefaultTabBar
				underlineStyle={styles.tabUnderlineStyle}
				activeTextColor={colors.primary}
				inactiveTextColor={colors.fontTertiary}
				backgroundColor={colors.white}
				tabStyle={styles.tabStyle}
				textStyle={styles.textStyle}
			/>
		);
	}

	render = () => {
		const {
			navigation: {
				state: {
					params: { assetType }
				}
			},
			navigation
		} = this.props;
		return (
			<SafeAreaView style={styles.wrapper} testID={`add-${assetType}-screen`}>
				{assetType === 'token' ? (
					<ScrollableTabView renderTabBar={this.renderTabBar}>
						<SearchTokenAutocomplete
							navigation={navigation}
							tabLabel={strings('add_asset.search_token')}
							testID={'tab-search-token'}
						/>
						<AddCustomToken
							navigation={navigation}
							tabLabel={strings('add_asset.custom_token')}
							testID={'tab-add-custom-token'}
						/>
					</ScrollableTabView>
				) : (
					<AddCustomCollectible navigation={navigation} testID={'add-custom-collectible'} />
				)}
			</SafeAreaView>
		);
	};
}
