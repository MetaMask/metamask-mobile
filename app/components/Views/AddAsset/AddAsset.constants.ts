export const TOKEN = 'token';
export const TOKEN_TITLE = 'title';
export const NFT_TITLE = 'title_nft';
