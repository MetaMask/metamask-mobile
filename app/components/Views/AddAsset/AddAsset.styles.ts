import { StyleSheet } from 'react-native';
import { Theme } from '../../../util/theme/models';

const styleSheet = (params: { theme: Theme }) => {
  const { theme } = params;
  const { colors } = theme;
  return StyleSheet.create({
    base: {
      paddingHorizontal: 16,
    },
    wrapper: {
      flex: 1,
      backgroundColor: colors.background.default,
    },
    infoWrapper: {
      alignItems: 'center',
      marginTop: 10,
      paddingHorizontal: 16,
    },
    tabUnderlineStyle: {
      height: 2,
      backgroundColor: colors.primary.default,
    },
    tabBar: {
      borderColor: colors.border.muted,
    },
    tabStyle: {
      paddingBottom: 0,
      paddingVertical: 8,
    },
    textStyle: {
      textTransform: 'capitalize',
      ...params.theme.typography.sBodyMD,
      fontWeight: '500',
    },
  });
};
export default styleSheet;
