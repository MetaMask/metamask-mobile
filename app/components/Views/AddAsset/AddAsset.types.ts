export interface AddAssetParams {
  assetType: string;
  collectibleContract?: {
    address: string;
  };
}
