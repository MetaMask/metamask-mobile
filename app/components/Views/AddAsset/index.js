import React, { PureComponent } from 'react';
import { SafeAreaView, View, StyleSheet } from 'react-native';
import { fontStyles } from '../../../styles/common';
import { connect } from 'react-redux';
import DefaultTabBar from 'react-native-scrollable-tab-view/DefaultTabBar';
import AddCustomToken from '../../UI/AddCustomToken';
import SearchTokenAutocomplete from '../../UI/SearchTokenAutocomplete';
import ScrollableTabView from 'react-native-scrollable-tab-view';
import PropTypes from 'prop-types';
import { strings } from '../../../../locales/i18n';
import AddCustomCollectible from '../../UI/AddCustomCollectible';
import { getNetworkNavbarOptions } from '../../UI/Navbar';
import CollectibleDetectionModal from '../../UI/CollectibleDetectionModal';
import { isTokenDetectionSupportedForNetwork } from '@metamask/assets-controllers/dist/assetsUtil';
import { ThemeContext, mockTheme } from '../../../util/theme';
import { MAINNET } from '../../../constants/network';

const createStyles = (colors) =>
  StyleSheet.create({
    wrapper: {
      flex: 1,
      backgroundColor: colors.background.default,
    },
    infoWrapper: {
      alignItems: 'center',
      marginTop: 10,
    },
    tabUnderlineStyle: {
      height: 2,
      backgroundColor: colors.primary.default,
    },
    tabBar: {
      borderColor: colors.border.muted,
    },
    tabStyle: {
      paddingBottom: 0,
    },
    textStyle: {
      fontSize: 16,
      letterSpacing: 0.5,
      ...fontStyles.bold,
    },
  });

/**
 * PureComponent that provides ability to add assets.
 */
class AddAsset extends PureComponent {
  state = {
    address: '',
    symbol: '',
    decimals: '',
    dismissNftInfo: false,
  };

  static propTypes = {
    /**
    /* navigation object required to push new views
    */
    navigation: PropTypes.object,
    /**
     * Network type
     */
    networkType: PropTypes.string,
    /**
     * Chain id
     */
    chainId: PropTypes.string,
    /**
     * Object that represents the current route info like params passed to it
     */
    route: PropTypes.object,
    /**
     * Boolean to show if NFT detection is enabled
     */
    useNftDetection: PropTypes.bool,
  };

  updateNavBar = () => {
    const { navigation, route } = this.props;
    const colors = this.context.colors || mockTheme.colors;
    navigation.setOptions(
      getNetworkNavbarOptions(
        `add_asset.${
          route.params.assetType === 'token' ? 'title' : 'title_nft'
        }`,
        true,
        navigation,
        colors,
      ),
    );
  };

  componentDidMount = () => {
    this.updateNavBar();
  };

  componentDidUpdate = () => {
    this.updateNavBar();
  };

  renderTabBar() {
    const colors = this.context.colors || mockTheme.colors;
    const styles = createStyles(colors);

    return (
      <DefaultTabBar
        underlineStyle={styles.tabUnderlineStyle}
        activeTextColor={colors.primary.default}
        inactiveTextColor={colors.text.alternative}
        backgroundColor={colors.background.default}
        tabStyle={styles.tabStyle}
        textStyle={styles.textStyle}
        style={styles.tabBar}
      />
    );
  }

  dismissNftInfo = async () => {
    this.setState({ dismissNftInfo: true });
  };

  render = () => {
    const {
      route: {
        params: { assetType, collectibleContract },
      },
      navigation,
      chainId,
      useNftDetection,
      networkType,
    } = this.props;
    const { dismissNftInfo } = this.state;
    const isTokenDetectionSupported =
      isTokenDetectionSupportedForNetwork(chainId);
    const colors = this.context.colors || mockTheme.colors;
    const styles = createStyles(colors);

    return (
      <SafeAreaView style={styles.wrapper} testID={`add-${assetType}-screen`}>
        {networkType === MAINNET &&
          assetType !== 'token' &&
          !dismissNftInfo &&
          !useNftDetection && (
            <View style={styles.infoWrapper}>
              <CollectibleDetectionModal
                onDismiss={this.dismissNftInfo}
                navigation={navigation}
              />
            </View>
          )}
        {assetType === 'token' ? (
          <ScrollableTabView
            key={chainId}
            renderTabBar={() => this.renderTabBar()}
          >
            {isTokenDetectionSupported && (
              <SearchTokenAutocomplete
                navigation={navigation}
                tabLabel={strings('add_asset.search_token')}
                testID={'tab-search-token'}
              />
            )}
            <AddCustomToken
              navigation={navigation}
              tabLabel={strings('add_asset.custom_token')}
              testID={'tab-add-custom-token'}
              isTokenDetectionSupported={isTokenDetectionSupported}
            />
          </ScrollableTabView>
        ) : (
          <AddCustomCollectible
            navigation={navigation}
            collectibleContract={collectibleContract}
            testID={'add-custom-collectible'}
          />
        )}
      </SafeAreaView>
    );
  };
}

AddAsset.contextType = ThemeContext;

const mapStateToProps = (state) => ({
  networkType: state.engine.backgroundState.NetworkController.provider.type,
  chainId: state.engine.backgroundState.NetworkController.provider.chainId,
  useNftDetection:
    state.engine.backgroundState.PreferencesController.useNftDetection,
});

export default connect(mapStateToProps)(AddAsset);
