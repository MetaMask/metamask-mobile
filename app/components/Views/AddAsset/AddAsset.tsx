import React, { useCallback, useEffect, useMemo } from 'react';
import { SafeAreaView, View } from 'react-native';
import { useSelector } from 'react-redux';
import DefaultTabBar from 'react-native-scrollable-tab-view/DefaultTabBar';
import AddCustomToken from '../../UI/AddCustomToken';
import SearchTokenAutocomplete from '../../UI/SearchTokenAutocomplete';
import ScrollableTabView from 'react-native-scrollable-tab-view';
import { strings } from '../../../../locales/i18n';
import AddCustomCollectible from '../../UI/AddCustomCollectible';
import {
  getImportTokenNavbarOptions,
  getNetworkNavbarOptions,
} from '../../UI/Navbar';
import { isTokenDetectionSupportedForNetwork } from '@metamask/assets-controllers/dist/assetsUtil';
import {
  selectChainId,
  selectProviderConfig,
} from '../../../selectors/networkController';
import { selectDisplayNftMedia } from '../../../selectors/preferencesController';
import Banner from '../../../component-library/components/Banners/Banner/Banner';
import {
  BannerAlertSeverity,
  BannerVariant,
} from '../../../component-library/components/Banners/Banner';
import Text from '../../../component-library/components/Texts/Text/Text';
import { TextVariant } from '../../../component-library/components/Texts/Text';
import { ButtonVariants } from '../../../component-library/components/Buttons/Button';
import { useNavigation } from '@react-navigation/native';
import { useParams } from '../../../util/navigation/navUtils';
import { useStyles } from '../../../component-library/hooks';
import styleSheet from './AddAsset.styles';
import { AddAssetParams } from './AddAsset.types';
import Routes from '../../../constants/navigation/Routes';
import { NFT_TITLE, TOKEN, TOKEN_TITLE } from './AddAsset.constants';
import { AddAssetViewSelectorsIDs } from '../../../../e2e/selectors/AddAssetView.selectors';
import { getNetworkNameFromProviderConfig } from '../../../util/networks';
const AddAsset = () => {
  const navigation = useNavigation();
  const { assetType, collectibleContract } = useParams<AddAssetParams>();

  const {
    styles,
    theme: { colors },
  } = useStyles(styleSheet, {});

  const providerConfig = useSelector(selectProviderConfig);
  const chainId = useSelector(selectChainId);
  const displayNftMedia = useSelector(selectDisplayNftMedia);

  const isTokenDetectionSupported =
    isTokenDetectionSupportedForNetwork(chainId);

  const networkName = useMemo(
    () => getNetworkNameFromProviderConfig(providerConfig),
    [providerConfig],
  );

  const updateNavBar = useCallback(() => {
    navigation.setOptions(
      assetType === TOKEN
        ? getImportTokenNavbarOptions(
            `add_asset.${TOKEN_TITLE}`,
            true,
            navigation,
            colors,
            true,
            0,
          )
        : getNetworkNavbarOptions(
            `add_asset.${assetType === TOKEN ? TOKEN_TITLE : NFT_TITLE}`,
            true,
            navigation,
            colors,
          ),
    );
  }, [assetType, colors, navigation]);

  useEffect(() => {
    updateNavBar();
  }, [updateNavBar]);

  const goToSecuritySettings = () => {
    navigation.navigate(Routes.MODAL.ROOT_MODAL_FLOW, {
      screen: Routes.SHEET.SHOW_NFT_DISPLAY_MEDIA,
    });
  };

  const renderTabBar = (props) => (
    <View style={styles.base}>
      <DefaultTabBar
        underlineStyle={styles.tabUnderlineStyle}
        activeTextColor={colors.primary.default}
        inactiveTextColor={colors.text.alternative}
        backgroundColor={colors.background.default}
        tabStyle={styles.tabStyle}
        textStyle={styles.textStyle}
        tabPadding={32}
        style={styles.tabBar}
        {...props}
      />
    </View>
  );

  return (
    <SafeAreaView style={styles.wrapper} testID={`add-${assetType}-screen`}>
      {assetType !== 'token' && (
        <View style={styles.infoWrapper}>
          <Banner
            variant={BannerVariant.Alert}
            description={
              !displayNftMedia ? (
                <>
                  <Text variant={TextVariant.BodyMD}>
                    {strings('wallet.display_nft_media_cta_new_1')}
                    <Text variant={TextVariant.BodyMDBold}>
                      {' '}
                      {strings('wallet.display_nft_media_cta_new_2')}
                    </Text>
                  </Text>
                </>
              ) : (
                <Text
                  variant={TextVariant.BodyMD}
                  testID={AddAssetViewSelectorsIDs.WARNING_ENABLE_DISPLAY_MEDIA}
                >
                  {strings('wallet.display_media_nft_warning')}
                </Text>
              )
            }
            severity={
              !displayNftMedia
                ? BannerAlertSeverity.Info
                : BannerAlertSeverity.Warning
            }
            actionButtonProps={
              !displayNftMedia
                ? {
                    variant: ButtonVariants.Link,
                    onPress: goToSecuritySettings,
                    textVariant: TextVariant.BodyMD,
                    label: strings('wallet.display_nft_media_cta'),
                  }
                : undefined
            }
          />
        </View>
      )}
      {assetType === 'token' ? (
        <ScrollableTabView key={chainId} renderTabBar={renderTabBar}>
          {isTokenDetectionSupported && (
            <SearchTokenAutocomplete
              navigation={navigation}
              tabLabel={strings('add_asset.search_token')}
            />
          )}
          <AddCustomToken
            chainId={chainId}
            networkName={networkName}
            ticker={providerConfig.ticker}
            type={providerConfig.type}
            navigation={navigation}
            tabLabel={strings('add_asset.custom_token')}
            isTokenDetectionSupported={isTokenDetectionSupported}
          />
        </ScrollableTabView>
      ) : (
        <AddCustomCollectible
          navigation={navigation}
          collectibleContract={collectibleContract}
        />
      )}
    </SafeAreaView>
  );
};

export default AddAsset;
