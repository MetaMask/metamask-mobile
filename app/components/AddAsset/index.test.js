import React from 'react';
import { shallow } from 'enzyme';
import AddAsset from './';

describe('AddAsset', () => {
	it('should render correctly', () => {
		const wrapper = shallow(<AddAsset />);
		expect(wrapper).toMatchSnapshot();
	});
});
