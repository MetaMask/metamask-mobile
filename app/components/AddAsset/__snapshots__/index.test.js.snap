// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddAsset should render correctly 1`] = `
<Component
  style={
    Object {
      "backgroundColor": "#dfe4ea",
      "flex": 1,
    }
  }
  testID="add-asset-screen"
>
  <ScrollableTabView
    contentProps={Object {}}
    initialPage={0}
    locked={false}
    onChangeTab={[Function]}
    onScroll={[Function]}
    page={-1}
    prerenderingSiblingsNumber={0}
    renderTabBar={[Function]}
    scrollWithoutAnimation={false}
    tabBarPosition="top"
  >
    <SearchAssetAutocomplete
      tabLabel="SEARCH"
      testID="tab-search-token"
    />
    <AddCustomAsset
      tabLabel="CUSTOM TOKEN"
      testID="tab-add-custom-token"
    />
  </ScrollableTabView>
</Component>
`;
